collector/mesos_agent/mesos_agent.go

```diff
@@ -174,3 +280,69 @@ func getLabelsByContainerID(containerID string, frameworks []frameworkInfo) map[
 	}
 	return labels
 }
+
+// buildDatapoints takes an incoming structure and builds Datapoints
+// for a MetricsMessage. It uses a normalized version of the JSON tag
+// as the datapoint name.
+func buildDatapoints(in interface{}, t time.Time) []producers.Datapoint {
+	pts := []producers.Datapoint{}
+	v := reflect.Indirect(reflect.ValueOf(in))
+
+	for i := 0; i < v.NumField(); i++ { // Iterate over fields in the struct
+		f := v.Field(i)
+		typ := v.Type().Field(i)
+
+		switch f.Kind() { // Handle nested data
+		case reflect.Ptr:
+			pts = append(pts, buildDatapoints(f.Elem().Interface(), t)...)
+			continue
+		case reflect.Map:
+			// Ignore maps when building datapoints
+			continue
+		case reflect.Slice:
+			for j := 0; j < f.Len(); j++ {
+				for _, ndp := range buildDatapoints(f.Index(j).Interface(), t) {
+					pts = append(pts, ndp)
+				}
+			}
+			continue
+		case reflect.Struct:
+			pts = append(pts, buildDatapoints(f.Interface(), t)...)
+			continue
+		}
+
+		// Get the underlying value; see https://golang.org/pkg/reflect/#Kind
+		var uv interface{}
+		switch f.Kind() {
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			uv = f.Int()
```

Comment:
I don't want to touch that code in this refactor, I'm not really sure how it works yet. Ok to leave it for this PR?